zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -137,15 +158,74 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
         );
 
         // Transfer value from caller to callee
+        let intrinsic_tx_value = cb.query_word();
+        let intrinsic_mul_gas_fee_by_gas = cb.query_word();
+        let intrinsic_tx_value_and_gas_fee_is_zero = BatchedIsZeroGadget::construct(
+            cb,
+            [
+                intrinsic_tx_value.expr(),
+                intrinsic_mul_gas_fee_by_gas.expr(),
+            ],
+        );
+        cb.condition(is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == 0 && intrinsic_mul_gas_fee_by_gas == 0",
+                intrinsic_tx_value_and_gas_fee_is_zero.expr(),
+                true.expr(),
+            );
+        });
+
+        cb.condition(1.expr() - is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_tx_value.expr(),
+                tx_value.expr(),
+            );
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_mul_gas_fee_by_gas.expr(),
+                mul_gas_fee_by_gas.product().expr(),
+            );
+        });
```

Comment:
This is done like this because `TransferWithGasFeeGadget` expects words instead of an expression, right? If so please add a comment because otherwise this looks a bit strange. :)

Reply:
Added